### PR TITLE
esp32: Update ESP-IDF to v5.2.2, optimise size

### PR DESCRIPTION
**Pull request by projectgus**
### Summary

This PR updates the esp32 port's ESP-IDF version to v5.2.2 and changes the version build in CI from v5.0.4 to v5.2.2.

This is in anticipation of adding support for new SoCs ESP32-C6 (see #11869) and ESP32-C2 (see #15440) in the linked PRs.

Propose after merging this and changing the nightly version then we wait a couple of weeks and if no regressions are reported then remove support for older ESP-IDF, including some code cleanup to suit.

### Code Size

The usual problem with updating ESP-IDF is an increase in code size. This PR applies some config changes to reduce resource usage, and also adds a utility script to quickly measure the static resource usage of multiple boards & ESP-IDF versions.

Current master branch resource usage:

| BOARD            | BOARD_VARIANT | IDF Version | Binary Size | Static IRAM Size | Static DRAM Size |
|------------------|---------------|-------------|-------------|------------------|------------------|
| ESP32_GENERIC    |               | v5.0.4      |     1669037 |           128286 |            52892 |
|                  |               | v5.0.5      |     1682469 |           125822 |            53732 |
|                  |               | v5.1.2      |     1708393 |           129398 |            54276 |
|                  |               | v5.2        |     1707709 |           131150 |            55588 |
|                  |               | v5.2.2      |     1663573 |           121966 |            55236 |
| ESP32_GENERIC    | D2WD          | v5.0.4      |     1425121 |           121270 |            52768 |
|                  |               | v5.0.5      |     1435673 |           118666 |            53608 |
|                  |               | v5.1.2      |     1456773 |           121306 |            54160 |
|                  |               | v5.2        |     1446181 |           121578 |            54776 |
|                  |               | v5.2.2      |     1399309 |           112318 |            54340 |
| ESP32_GENERIC    | SPIRAM        | v5.0.4      |     1545837 |           130618 |            53308 |
|                  |               | v5.0.5      |     1557441 |           127802 |            54124 |
|                  |               | v5.1.2      |     1551761 |           128770 |            54540 |
|                  |               | v5.2        |     1543413 |           126870 |            55208 |
|                  |               | v5.2.2      |     1497321 |           117170 |            54788 |
| ESP32_GENERIC_S3 |               | v5.0.4      |     1564905 |           108674 |            49080 |
|                  |               | v5.0.5      |     1581793 |           104290 |            49344 |
|                  |               | v5.1.2      |     1595049 |            16383 |           138071 |
|                  |               | v5.2        |     1622129 |            16383 |           140159 |
|                  |               | v5.2.2      |     1632013 |            16383 |           143315 |
| ESP32_GENERIC_S3 | SPIRAM_OCT    | v5.0.4      |     1568157 |           110966 |            49980 |
|                  |               | v5.0.5      |     1585081 |           106582 |            50236 |
|                  |               | v5.1.2      |     1598289 |            16383 |           141267 |
|                  |               | v5.2        |     1625693 |            16383 |           143743 |
|                  |               | v5.2.2      |     1635609 |            16383 |           146923 |

Note that the ESP32-S3 pools IRAM and DRAM, so the last two columns need to be added together to be meaningful.

Resource usage after applying changes in this PR:

 BOARD | BOARD_VARIANT | IDF Version | Binary Size | Static IRAM Size | Static DRAM Size |
|-------|---------------|-------------|-------------|------------------|------------------|
| ESP32_GENERIC |  | v5.2.2 | 1627321 | 115310 | 55236 |
| ESP32_GENERIC | D2WD | v5.2.2 | 1361777 | 108074 | 54332 |
| ESP32_GENERIC | SPIRAM | v5.2.2 | 1459885 | 113046 | 54780 |
| ESP32_GENERIC_S3 |  | v5.2.2 | 1574649 | 16383 | 135299 |
| ESP32_GENERIC_S3 | SPIRAM_OCT | v5.2.2 | 1578245 | 16383 | 138907 |

Summary:

* For ESP32. the total binary size and the IRAM usage is now *smaller* on v5.2.2 with these changes compared to master branch with ESP-IDF v5.0.4.
* For ESP32-S3 the total binary size got bigger moving from v5.0.4 to v5.2.2(!) but after applying these changes it's  only a little bigger than ESP-IDF v5.0.4. Not sure why this is...
* About 2344 less bytes of DRAM is available at runtime for ESP-IDF heap, compared to 5.0.4. As far as I can tell this is because of more "iram safe" error log strings that are linked to static DRAM. I can't see any easy way to tackle this, unfortunately.

### Testing

* [x] Run unit tests on ESP32 and ESP32-S3
* [x] Add new unit test for SPI in linked PR
* [x] Run multi_bluetooth and multi_net tests on ESP32 and ESP32-S3
* [x] Test ESP32-C3 with built-in USB device

### Other TODOs

* [x] Resolve mbedTLS 3.6 upgrade fix
* [x] Merge stack check fix for ESP32-C3 https://github.com/micropython/micropython/pull/15605

### Trade-offs and Alternatives

The main trade-off decision is keeping vs dropping support for older ESP-IDF, particularly as there is a small impact to available DRAM on v5.2.2. However supporting three different minor versions (5.0, 5.1, 5.2) is a lot of additional overhead for MicroPython, and the difference between 5.1 and 5.2 doesn't seem like it really justifies keeping 5.1 support around. Let me know if I'm wrong!

### CODE REVIEW

**Comment by projectgus** [comment]
Rebased to pick up the changes in #15476, the numbers in the main post will be slightly wrong now.

**Comment by projectgus** [comment]
I've started a discussion on Discord about more completely dropping support for older ESP-IDF in this PR, as there are a bunch of compatibility checks that we can simplify out:
https://discord.com/channels/574275045187125269/952412628926472273/1265452441932136550

Comments welcome.

**Comment by projectgus** [comment]
Have run unit tests, multi_bluetooth and multi_net (Wi-Fi) tests on ESP32 and ESP32-S3. Found one issue introduced by updating from v5.0.4 to v5.2.2 (mbedTLS-related commit in this PR), plus several other broken tests or bugs that were pre-existing (to be submitted in separate PRs).

With all these changes present I now see no test failures, apart from some quite infrequent stress test failures. :tada: 

I wasn't able to test on ESP32-C3, both the boards I have are quite old (probably predating the C3's public release) and don't boot reliably on v5.2.2. I see some debug logs about missing eFuse calibration data, so I assume this is an issue with my boards and not MicroPython. Have ordered a newer C3 board, but if someone has a C3 at hand do they mind checking that it boots OK please?

@dpgeorge Based on @andrewleech's observation in Discord that some people ran into Wi-Fi issues with v5.2, I suggest we keep all versions supported in this PR but switch CI and nightly builds to v5.2.2. If no significant reports are received in a couple of weeks then I'll submit another PR to drop support for older versions and cleanup the code to suit. What do you think?

**Comment by dpgeorge** [comment]
> if someone has a C3 at hand do they mind checking that it boots OK please

I will check a C3.

> I suggest we keep all versions supported in this PR but switch CI and nightly builds to v5.2.2. If no significant reports are received in a couple of weeks then I'll submit another PR to drop support for older versions and cleanup the code to suit. What do you think?

Sounds good!

**Comment by andrewleech** [comment]
I've just compiled and flashed this to a new C3 board (recently received, been meaning to try it) [using official IDF v5.2.2 docker](https://github.com/micropython/micropython/wiki/Compiling-esp32-port-with-docker)

```
anl@STEP: ~/micropython $ esp-compile v5.2.2 ESP32_GENERIC_C3 /dev/ttyACM0
... snip ...
Wrote 3072 bytes at 0x00008000 in 0.0 seconds (599.7 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
Done
make: Leaving directory '/home/anl/micropython/ports/esp32'

anl@STEP: ~/micropython $ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell

>>>
MPY: soft reboot
MicroPython v1.24.0-preview.137.gec63829b07.dirty on 2024-07-25; ESP32C3 module with ESP32C3
Type "help()" for more information.
>>>
```

**Comment by andrewleech** [comment]
I just ran multi tests between a C3 and S3 (FLASH_4M), both running 5.2.2 @ ec63829b07:
```
anl@STEP: ~/micropython/tests $ ./run-multitests.py -ipyb:/dev/ttyACM0 -ipyb:/dev/ttyACM1 multi_net/*.py
multi_net/asyncio_tcp_client_rst.py on ttyACM0|ttyACM1: skip
multi_net/asyncio_tcp_close_write.py on ttyACM0|ttyACM1: pass
multi_net/asyncio_tcp_readall.py on ttyACM0|ttyACM1: pass
multi_net/asyncio_tcp_readexactly.py on ttyACM0|ttyACM1: pass
multi_net/asyncio_tcp_readinto.py on ttyACM0|ttyACM1: pass
multi_net/asyncio_tcp_server_client.py on ttyACM0|ttyACM1: pass
multi_net/asyncio_tls_server_client.py on ttyACM0|ttyACM1: skip
multi_net/asyncio_tls_server_client_cert_required_error.py on ttyACM0|ttyACM1: skip
multi_net/asyncio_tls_server_client_readline.py on ttyACM0|ttyACM1: skip
multi_net/asyncio_tls_server_client_verify_error.py on ttyACM0|ttyACM1: skip
multi_net/ssl_cert_ec.py on ttyACM0|ttyACM1: skip
multi_net/ssl_cert_rsa.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_check_hostname_error.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_getpeercert.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_server_client.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_server_client_ciphers.py on ttyACM0|ttyACM1: FAIL
### TEST ###
--- instance0 ---

Traceback (most recent call last):
  File "<stdin>", line 16, in <module>
OSError: [Errno 2] ENOENT

--- instance1 ---

### TRUTH ###
--- instance0 ---
b'client to server'
--- instance1 ---
b'server to client'
### DIFF ###
--- /tmp/tmp5czsrt2w    2024-07-25 16:09:40.301553122 +1000
+++ /tmp/tmpgz1a7lb_    2024-07-25 16:09:40.301553122 +1000
@@ -1,4 +1,8 @@
 --- instance0 ---
-b'client to server'
+
+Traceback (most recent call last):
+  File "<stdin>", line 16, in <module>
+OSError: [Errno 2] ENOENT
+
 --- instance1 ---
-b'server to client'
+
multi_net/sslcontext_server_client_files.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_verify_callback.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_verify_error.py on ttyACM0|ttyACM1: skip
multi_net/sslcontext_verify_time_error.py on ttyACM0|ttyACM1: skip
multi_net/tcp_accept_recv.py on ttyACM0|ttyACM1: pass
multi_net/tcp_client_rst.py on ttyACM0|ttyACM1: skip
multi_net/tcp_data.py on ttyACM0|ttyACM1: pass
multi_net/udp_data.py on ttyACM0|ttyACM1: pass
24 tests performed
8 tests passed
15 tests skipped: multi_net/asyncio_tcp_client_rst.py multi_net/asyncio_tls_server_client.py multi_net/asyncio_tls_server_client_cert_required_error.py multi_net/asyncio_tls_server_client_readline.py multi_net/asyncio_tls_server_client_verify_error.py multi_net/ssl_cert_ec.py multi_net/ssl_cert_rsa.py multi_net/sslcontext_check_hostname_error.py multi_net/sslcontext_getpeercert.py multi_net/sslcontext_server_client.py multi_net/sslcontext_server_client_files.py multi_net/sslcontext_verify_callback.py multi_net/sslcontext_verify_error.py multi_net/sslcontext_verify_time_error.py multi_net/tcp_client_rst.py
1 tests failed: multi_net/sslcontext_server_client_ciphers.py
```

Ah I see that failed test requires extra files created/copied onto the devices, my bad

**Comment by dpgeorge** [comment]
> I'm guessing you need to set the time on the devices somehow

Indeed!  Try `mpremote rtc --set`.

**Comment by andrewleech** [comment]
> 
> Indeed! Try `mpremote rtc --set`.

Oh that's easier than I was guessing, I forgot about that command thanks!
I've updated my readme pr with an appropriate note.

File: extmod/modtls_mbedtls.c [hunk 1]
```diff
@@ -344,24 +344,41 @@ static mp_obj_t ssl_context_set_ciphers(mp_obj_t self_in, mp_obj_t ciphersuite)
 }
 static MP_DEFINE_CONST_FUN_OBJ_2(ssl_context_set_ciphers_obj, ssl_context_set_ciphers);
 
+// Given a string-like object, return a ASN.1 container buffer (PEM or DER) and its
+// correct length. Result is not necessarily a C string.
+//
+// Starting with mbedTLS 3.5, DER formatted keys must consume the entire keylen
+// argument (no trailing bytes) or parsing fails. However, PEM formatted keys
+// must be submitted with a terminating NUL byte included in the keylen...
```
Comment (dpgeorge) [inline extmod/modtls_mbedtls.c:352]: This fix is actually being worked on separately in #14385.

**Comment by dpgeorge** [comment]
I tested this PR with the two C3 boards that I have:
- ESP32-C3-DevKitC-02
- ESP32-C3-DevKitM-1

And IDF 5.2.2 with this PR does not work on either of them.

Looking at the DevKitC in particular, which has a ESP32-C3-WROOM-02 module.  With this PR and IDF 5.2.2 the output on the /dev/ttyUSB0 serial when I press the RST button is:
```
$ mpremote
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5820,len:0xf28
load:0x403cc710,len:0x89c
load:0x403ce710,len:0x2b14
entry 0x403cc710
```

When I use this PR but with IDF 5.0.5 I get:
```
$ mpremote
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5820,len:0xe8c
load:0x403cc710,len:0x6ec
load:0x403ce710,len:0x2b24
entry 0x403cc710
MicroPython v1.24.0-preview.155.g768bdf99d on 2024-07-26; ESP32C3 module with ESP32C3
Type "help()" for more information.
>>> 
```

These are pretty old boards, but I think they should still work with IDF 5.2.2, considering they work with IDF 5.0.5.  Comparing to @andrewleech 's C3 test above:
- the DevKitC takes about twice as long to download firmware: Wrote 1602560 bytes at 0x00010000 in 56.3 seconds (227.6 kbit/s)...
- it uses /dev/ttyUSB0 (not /dev/ttyACM0)

**Comment by andrewleech** [comment]
How interesting! With DevKitC is that via a UART bridge?

My board is a cheap ESP32 C3 SuperMini which only exposes the built-in usb, this drops out on rst button and comes up as ACM.

**Comment by dpgeorge** [comment]
There must be a UART-USB bridge, because the USB stays alive over a reset.

The flashing output for this board is:
```
esptool.py --chip esp32c3 -p /dev/ttyUSB0 -b 460800 --before=default_reset --after=hard_reset --no-stub write_flash --flash_mode dio --flash_freq 80m --flash_size 4MB 0x0 bootloader/bootloader.bin 0x10000 micropython.bin 0x8000 partition_table/partition-table.bin
esptool.py v4.7.0
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP32-C3 (QFN32) (revision v0.3)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 7c:df:a1:af:4b:4c
Changing baud rate to 460800
Changed.
Enabling default SPI flash mode...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00004fff...
Flash will be erased from 0x00010000 to 0x001a3fff...
Flash will be erased from 0x00008000 to 0x00008fff...
...
```

**Comment by projectgus** [comment]
> These are pretty old boards, but I think they should still work with IDF 5.2.2, considering they work with IDF 5.0.5.

Agree. These are the same symptoms mine have, so I guess it's more widespread than just early silicon. Will look into it more.

**Comment by dpgeorge** [comment]
> Agree. These are the same symptoms mine have

Mine don't give any errors about efuse.

Do your ESP32C3's work with this PR and IDF 5.0.5?

**Comment by projectgus** [comment]
> Mine don't give any errors about efuse.

They're logged at Debug level so you have to mess with the IDF config to see them. It's possible all of the chips from this revision (ECO3) have the same missing efuses.

Mine work with IDF v5.0.4, I haven't tried v5.0.5. The other weird thing is that they work with the IDF v5.2.2 hello_world example and the default sdkconfig. So it's probably a config flag in the MicroPython config, unexpectedly.

**Comment by andrewleech** [comment]
Maybe on newer IDF the handling of this being blank is to disable the uart0 output? https://github.com/micropython/micropython/blob/ec63829b070cbeb132803ee8a787c062837ef7cb/ports/esp32/boards/ESP32_GENERIC_C3/sdkconfig.c3usb#L6

On the C6 discussion there was chatter about needing  to add the baudrate too, though that was to allow it to build https://github.com/micropython/micropython/pull/11869#issuecomment-1826180807

Based on some reports there I think the uart repl might be not working out of the box in C6 too, possibly the same issue here: https://github.com/micropython/micropython/pull/11869#issuecomment-2057017836

**Comment by projectgus** [comment]
> Maybe on newer IDF the handling of this being blank is to disable the uart0 output?

@andrewleech yes, this was it, thanks! Nothing to do with silicon revisions. :sweat_smile: 

I have pushed a commit a2de9c7 that works with both v5.0.4 and v5.2.2 on the UART console, by going back to the default ESP-IDF console config. More details in the commit message. I don't have a board with the native USB serial/JTAG port exposed (one more that's on the way), if you have time could you test it on yours please?

**Comment by dpgeorge** [comment]
I've tested the latest version of this PR and it now works on my ESP32-C3-DevKitC-02 board.

Running all the tests, one fails: `thread/stress_recurse.py`.  It looks like a stack overflow.  This tests passes with this PR but using IDF 5.0.5.  So I guess that IDF 5.2.2 somehow uses a little more stack for threads.

**Comment by dpgeorge** [comment]
> Running all the tests, one fails: `thread/stress_recurse.py`. It looks like a stack overflow. This tests passes with this PR but using IDF 5.0.5. So I guess that IDF 5.2.2 somehow uses a little more stack for threads

See #15583 for a proposed fix for that.  @projectgus you may want to cherry-pick that commit into this PR to make testing easier.

**Comment by projectgus** [comment]
Rebased.

My "ESP32-C3 SuperMini" board turned up, using the integrated Serial/JTAG. It seems to work correctly with the serial console config changes in this PR, also. Unit tests pass apart from:

* math_domain_special (pre-existing, fix in #15539).
* stress_recurse (possibly pre-existing possibly a 5.2.2 regression, fix in #15605).

* Re-ran multi_bluetooth tests, they pass on S3 and C3[*]
* Re-ran multi_net tests, also pass on S3 and C3.

[*]  *Except* my new SuperMini board randomly resets sometimes mid-test for Bluetooth. Espressif DevKit-C3 does not do this. I think the Brownout Detector is triggering, it's an unbelievably cheap board and doesn't look to have a lot of capacitors on it. This isn't a new issue, I see the same on master with V5.0.4 - so I don't think it's a blocker for this PR.

@dpgeorge This is ready apart from #15605 and also the mbedTLS certificate parsing fix. I don't think we should change the default IDF version to V5.2.2 without the mbedTLS fix, as cert parsing will fail. I still have a fix commit for that in this branch (86692a515dc62812e618f3834f2bd3dbb4b759af), but I also made a suggestion to the author of the other fix PR here: https://github.com/micropython/micropython/pull/14385#discussion_r1697718241 . Do you have a preference for how to proceed on that one?

**Comment by projectgus** [comment]
(Rebased again, cleaned up the metrics_esp32.py script and updated the numbers in the PR description.)

**Comment by dpgeorge** [comment]
The mbedtls and stack-check PRs are now merged, so this is ready for a (final?) rebase and review.

**Comment by projectgus** [comment]
@dpgeorge Thanks! Rebased, re-ran unit tests on ESP32, C3, and S3. I didn't rerun additional tests as I don't think there's a high risk of regressions since last test run.

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
I tested this on a ESP32-C3-DevKitC-02, running the main test suite, and then the network and bluetooth multitests.  Everything passes (except `multi_bluetooth/ble_irq_calls.py` but that's currently broken on master).

**Comment by dpgeorge** [comment]
Now merged!

Thanks @projectgus for all your effort testing this.

**Pull request merged**